### PR TITLE
[#838] ZoomTab: add sort controls 

### LINKED ISSUE #838: Add sorting options for zoom tab when viewing commits
Current: Zoom tab shows the list of commits in the order from earliest to latest commits (without an option to reverse the sorting order).

image

Suggested: -
- Let users see the list of commits in **reverse chronological order**, and/or
- Add other sorting options? Such as Sorting commits by the number of LoC added/modified in each commit or each day?

### PR DESCRIPTION
**Pull request by 0blivious**
Fixes #838
```
We have Zoom tab shows the list of commits in the order from
earliest to latest commits (without an option to reverse the sorting
order or change the sorting type).

This can be improved by allowing the user to view the commits in
reverse chronological order and add other sorting options such as
LOC.

Let's add a sorting field to allow the user to sort the commits
according to Loc or time, and also allow the commits to be
displayed in reverse order.
```

### CODE REVIEW

**Comment by jamessspanggg** [comment]
@damithc since we are default sorting code view in terms of descending order of LoC, should we do the same for zoom view as well? i.e. default as time/LoC with reverse order

**Comment by damithc** [comment]
> @damithc since we are default sorting code view in terms of descending order of LoC, should we do the same for zoom view as well? i.e. default as time/LoC with reverse order

Is there a sorting order for that view?

**Comment by 0blivious** [comment]
> > @damithc since we are default sorting code view in terms of descending order of LoC, should we do the same for zoom view as well? i.e. default as time/LoC with reverse order
> 
> Is there a sorting order for that view?

This pr aims to add a sorting option for the zoom tab as mentioned in the issue #838

**Comment by damithc** [comment]
> This pr aims to add a sorting option for the zoom tab as mentioned in the issue #838

Got it. I guess it makes sense to show the latest commits at the top by default?

**Comment by 0blivious** [comment]
> Got it. I guess it makes sense to show the latest commits at the top by default?

Then for the LOC sorting we also display the large number one at top can? I feel this makes the sorting more coherent as we always sort by descending order.

**Comment by jamessspanggg** [comment]
> > Got it. I guess it makes sense to show the latest commits at the top by default?
> 
> Then for the LOC sorting we also display the large number one at top can? I feel this makes the sorting more coherent as we always sort by descending order.

Yes I think it is better to show the LoC to be in descending order as the default, if we are making the most recent commits to be at the top.

**Comment by 0blivious** [comment]
> Yes I think it is better to show the LoC to be in descending order as the default, if we are making the most recent commits to be at the top.

The previous PR has already made both sorting by time and by LOC in descending order.

**Comment by Tejas2805** [comment]
> The previous PR has already made both sorting by time and by LOC in descending order.

You might want to check your Netlify preview regarding this.

**Comment by 0blivious** [comment]
> You might want to check your Netlify preview regarding this.

Fixed

**Comment by fzdy1914** [comment]
> > Yes I think it is better to show the LoC to be in descending order as the default, if we are making the most recent commits to be at the top.
> 
> The previous PR has already made both sorting by time and by LOC in descending order.

To standardize with another ongoing PR #932, should we make the `reversed` checkbox as a drop-down as well?

**Comment by Tejas2805** [comment]
> To standardize with another ongoing PR #932, should we make the `reversed` checkbox as a drop-down as well?

I feel the current UI looks neat, but the checkbox being checked from before looks weird. 

Also, if we do go for dropdown we will need to maintain consistency between `sort by` and `order`

**Comment by 0blivious** [comment]
> > To standardize with another ongoing PR #932, should we make the `reversed` checkbox as a drop-down as well?
> 
> I feel the current UI looks neat, but the checkbox being checked from before looks weird.
> 
> Also, if we do go for dropdown we will need to maintain consistency between `sort by` and `order`

Or I can follow the sorting in summary panel? Can put the order inside the sorting drop down so we have sth like sort by `time ↑` and `time ↓`. I think don't need an extra drop down menu just for asc and des order.

**Comment by Tejas2805** [comment]
Based on @eugenepeh comment on PR #948 we will be trying to make the two views similar to each other. Let's for him to respond

**Comment by eugenepeh** [comment]
>Based on @eugenepeh comment on PR #948 we will be trying to make the two views similar to each other. Let's wait for him to respond

Yes, for consistency, please reuse components from other views as much as possible

**Comment by 0blivious** [comment]
Sorting option changed to drop down menu

File: frontend/src/static/js/v_zoom.js [hunk 1]
```diff
@@ -1,3 +1,8 @@
+const commitSortDict = {
+  lineOfCode: (commit) => commit.insertions,
```
Comment (fzdy1914) [inline frontend/src/static/js/v_zoom.js]: why this dict can not be a data of `vZoom`?

File: frontend/src/tabs/zoom.pug [hunk 2]
```diff
@@ -21,11 +21,23 @@
       v-bind:avgsize="info.avgCommitSize",
       v-bind:mergegroup="info.isMergeGroup")
 
-  .zoom__toggle-commit-message-body(v-if="info.user.commits.length")
-    button(v-if="expandedCommitMessagesCount === 0" v-on:click="toggleAllCommitMessagesBody(true)") show all commit messages body
-    button(v-else v-on:click="toggleAllCommitMessagesBody(false)") hide all commit messages body
+  .zoom__toolbar
+      .toggle-commit-message-body(v-if="info.user.commits.length")
```
Comment (fzdy1914) [inline frontend/src/tabs/zoom.pug]: Indentation.

File: frontend/src/static/js/v_zoom.js (hunk 1 above)
Comment (0blivious) [reply frontend/src/static/js/v_zoom.js]: Vue sets all the properties in the data to setters/getters to make them reactive and I feel this dict should be static.

Comment (fzdy1914) [reply frontend/src/static/js/v_zoom.js]: OK, got what you mean.

**Comment by fzdy1914** [comment]
@0blivious Can you kindly resolve the conflicts?

File: frontend/src/static/js/v_zoom.js [hunk 3]
```diff
@@ -6,8 +11,22 @@ window.vZoom = {
       filterTimeFrame: window.hashParams.timeframe,
       showAllCommitMessageBody: true,
       expandedCommitMessagesCount: this.totalCommitMessageBodyCount,
+      commitsSortType: 'time',
+      toReverseSortedCommits: true,
+      sortingFunction: (a, b) => -1 * window.comparator(commitSortDict.time)(a, b),
```
Comment (fzdy1914) [inline frontend/src/static/js/v_zoom.js]: Why not making `sortingFunction` a computed value? So you do no need to watch `commitsSortType` and `toReverseSortedCommits` anymore?

**Comment by fzdy1914** [comment]
@damithc , Can you help checking the UI?

**Comment by damithc** [comment]
> @damithc , Can you help checking the UI?

Yes 👍

File: frontend/src/static/css/style.scss [hunk 4]
```diff
@@ -685,8 +685,30 @@ header {
...
+      }
+
+      & > label {
+        margin-left: 10px;
+      }
+    }
+
+    .toggle-commit-message-body {
+      width: 50%;
+    }
```
Comment (eugenepeh) [inline frontend/src/static/css/style.scss]: Seems overkill given that we're reusing existing style. 
Can't we just standardize the both of them to be on the left side and save all this changes?
The `hide all commit messages body` can be changed to how `Collapse all` is like in authorship tab

Comment (0blivious) [reply frontend/src/static/css/style.scss]: For the `hide all commit messages body`, there is already a issue regarding this (#993 ). Can I check if you prefer tick box or the authorship style? If the authorship style, you want me to make the change in this PR?

Comment (eugenepeh) [reply frontend/src/static/css/style.scss]: Authorship style would be better, since its also a toggle button.
Can add it to this PR if its light changes.

**Comment by 0blivious** [comment]
> Since we are standardizing the style, the word choice should be the same either. Like `hide/show` and `collapse/expand` are not same. Meanwhile, in authorship, it should add, `expand file detail` or a similar word.
> 
> Other styles are OK for me.

Now I use hide/show file details for `authorship` and hide/show commit messages for `zoom`. Is this ok?

**Comment by fzdy1914** [comment]
> > Since we are standardizing the style, the word choice should be the same either. Like `hide/show` and `collapse/expand` are not same. Meanwhile, in authorship, it should add, `expand file detail` or a similar word.
> > Other styles are OK for me.
> 
> Now I use hide/show file details for `authorship` and hide/show commit messages for `zoom`. Is this ok?

Word `all` is missing which I think is necessary.

File: myconfig/repo-config.csv [hunk 5]
```diff
@@ -0,0 +1,2 @@
+Repository's Location,Branch,File formats,Ignore Glob List,Ignore standalone config,Ignore Commits List
```
Comment (fzdy1914) [inline myconfig/repo-config.csv]: Why this change?

Comment (fzdy1914) [reply myconfig/repo-config.csv]: Seems unrelated?

Comment (0blivious) [reply myconfig/repo-config.csv]: Sorry some error occured. This folder is supposed to be ignored

File: frontend/src/static/css/style.scss [hunk 6]
```diff
@@ -453,6 +454,34 @@ header {
   }
 }
 
+.toolbar {
```
Comment (fzdy1914) [inline frontend/src/static/css/style.scss]: Can you please place this whole peace at the end of same file. Because these lines will not be split into its own file anymore.

**Comment by 0blivious** [comment]
@jylee-git The first bug has been fixed.

> 2. Not sure if this is a bug or intended, but are you trying to make the sorting by LoC such that the commit that made the most contributions on that day will be taken into account?

This is intended. In `zoom panel` we provide date for each commit as the title. I feel if we break it down into commits the page will be a bit messy. Would you prefer we sort them by inividual LoC of each commits, if `granularity` is commit?

**Comment by jylee-git** [comment]
> This is intended. In `zoom panel` we provide date for each commit as the title. I feel if we break it down into commits the page will be a bit messy. Would you prefer we sort them by inividual LoC of each commits, if `granularity` is commit?

I'm personally fine with this, but not sure about what the others think. Will leave this message here for the others to read through

**Comment by fzdy1914** [comment]
> 1. Some commit messages that are initially hidden gets expanded. Do help to investigate this
> 

Can you look into this?

**Comment by 0blivious** [comment]
> > 1. Some commit messages that are initially hidden gets expanded. Do help to investigate this
> 
> Can you look into this?

It is already fixed in the latest commit

**Comment by 0blivious** [comment]
> Do update the screenshots and GIF file in the UG
Noted, I will update the report after we decide what to do with the commit display.

File: frontend/src/static/js/v_zoom.js [hunk 7]
```diff
@@ -6,17 +11,24 @@ window.vZoom = {
...
       expandedCommitMessagesCount: this.totalCommitMessageBodyCount,
+      commitsSortType: 'time',
+      toReverseSortedCommits: true,
     };
   },
+
   computed: {
+    sortingFunction() {
+      return (a, b) => (this.toReverseSortedCommits ? -1 : 1)
+      * window.comparator(commitSortDict[this.commitsSortType])(a, b);
```
Comment (Tejas2805) [inline frontend/src/static/js/v_zoom.js]: Can we make this change for authorship too? Will help in maintaining consistency.

Comment (fzdy1914) [reply frontend/src/static/js/v_zoom.js]: It can be done in another PR though.

Comment (Tejas2805) [reply frontend/src/static/js/v_zoom.js]: Okay sure

**Comment by Tejas2805** [comment]
@fzdy1914 Can merge?

File: frontend/src/static/css/style.scss [hunk 8]
```diff
@@ -299,6 +300,18 @@ header {
       }
     }
   }
+
+  .commit-message {
```
Comment (fzdy1914) [inline frontend/src/static/css/style.scss]: Seems we got two div with same `.commit-message` now? Why it is necessary?

Comment (0blivious) [reply frontend/src/static/css/style.scss]: removed
